### PR TITLE
Per Particle Logger

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

This PR adds per particle Hamiltonian reporting to estimators through the entire application and provides the proof of concept particle Hamiltonian logger. The logger is not optimized for performance nor is the output optimal for actual use.  If anyone would like to improve it further they should feel free.

More complete documentation of the Estimator architecture will be added in future PR's.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Build related changes
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

LECONTE

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes This PR is up to date with current the current state of 'develop'
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/Crowd.h [hunk 1]
```diff
@@ -104,6 +108,8 @@ class Crowd
   const MultiWalkerDispatchers& dispatchers_;
 
 private:
+  /// For some output purposes its useful if a crowd has an identifier for it on a rank.
+  const int crowd_id_ = 0;
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/Crowd.h:112]: I'm reconsidering this state, I thought it was necessary but now I think it is just lazy.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/Crowd.h]: I decided to eliminate the crowd_id_ and simply take the Walker.ID seriously.  This is a more likely desired for a per particle log anyway.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/Crowd.h]: Waker.ID is as bad as crowd_id_. It was not taken seriously.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/Crowd.h]: Well it requires less disturbance to the code. To log per particle per walker hamiltonian components there must be some identification of the walkers.  I added unit testing of the walker ids.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/Crowd.h]: walker ID is global (across all MPI ranks), very hard to keep track of. You can create some local data structure for the indexing, it is far better than replying on an global ID.

**Comment by prckent** [comment]
I would like to break the symmetry here and get this activity moving again, not only this PR. Peter and I discussed this last week.

First of all, since this PR doesn't break existing functionality, the general guidance about code moving in a good direction being mergeable applies. 

Second, one of the use cases for this level of logging is to help diagnose a failed run, at least on "replay" using the same seed and run configuration. In this case having a unique ID for each walker would be very useful, so this is what we should work towards. There is no question that we would have used the capability for this purpose several times in the last year.

In VMC the IDs are trivial, since the walker count is constant. In DMC we need e.g. an incrementing counter and to make sure we can count 100s millions of walkers. Since the walker creation/destruction logic is global this is also not difficult. A new ID is only needed when new walkers are created. The required state information is a single (long?) integer. This can be worked towards in subsequent PRs.

As with general use of the per particle logger, we are concerned with functionality and achievable insight, not straight line speed.

File: src/Estimators/EstimatorManagerCrowd.cpp [hunk 2]
```diff
@@ -38,6 +38,13 @@ void EstimatorManagerCrowd::accumulate(const RefVector<MCPWalker>& walkers,
     operator_ests_[i]->accumulate(walkers, psets, wfns, rng);
 }
 
+void EstimatorManagerCrowd::registerListeners(const RefVectorWithLeader<QMCHamiltonian>& ham_list)
+{
+  ///QMCHamiltonian& ham_leader = ham_list.getLeader();
```
Comment (ye-luo) [inline src/Estimators/EstimatorManagerCrowd.cpp:43]: Remove commented line

File: src/Particle/LongRange/StructFact.h [hunk 3]
```diff
@@ -48,6 +48,9 @@ class StructFact : public QMCTraits
   /** Constructor - copy ParticleSet and init. k-shells
    * @param lattice long range box
    * @param kc cutoff for k
+   *
+   * At least in the batched version Structure factor is _NOT_ valid
+   * after construction.
```
Comment (ye-luo) [inline src/Particle/LongRange/StructFact.h:53]: It seems you are trying to describe an issue. Could you put it as an issue in github?

File: src/QMCDrivers/Crowd.h [hunk 4]
```diff
@@ -44,6 +44,9 @@ class Crowd
    */
   Crowd(EstimatorManagerNew& emb,
         const DriverWalkerResourceCollection& driverwalker_res,
+	const ParticleSet& pset,
+        const TrialWaveFunction& twf,
+	const QMCHamiltonian& hamiltonian_temp,
```
Comment (ye-luo) [inline src/QMCDrivers/Crowd.h:49]: I feel very bad about adding this "gold" element here. Could you leave a comment about why they are added here. Probably need to get rid of them eventually.

File: src/QMCDrivers/QMCDriverNew.cpp [hunk 5]
```diff
@@ -153,19 +151,33 @@ void QMCDriverNew::initializeQMC(const AdjustedWalkerCounts& awc)
...
   {
+    // This needs to happen before walkers are made. i.e. this allows hamiltonian operators to update state
+    // the based on the presence of per particle listeners. In the case immediately encountered the operator CoulombPBCAA will
+    // call its associated particle set and turnOnPerParticleSK.
+    // The design for "initialization" of walker elements is for the golden elements to go through all pre walking state changes
+    // and then for the golden elements to be cloned for each walker.
+    if (estimator_manager_->areThereListeners())
+    {
+      population_.get_golden_hamiltonian().informOperatorsOfListener();
+    }
```
Comment (ye-luo) [inline src/QMCDrivers/QMCDriverNew.cpp:162]: Remove {} around one line.

File: src/Utilities/for_testing/NativeInitializerPrint.hpp [hunk 6]
```diff
@@ -72,7 +72,7 @@ std::ostream& operator<<(std::ostream& out, const NativePrint<Vector<T>>& np_vec
   return out;
 }
 
-
+
```
Comment (ye-luo) [inline src/Utilities/for_testing/NativeInitializerPrint.hpp:75]: Formatting.

**Comment by ye-luo** [comment]
> I would like to break the symmetry here and get this activity moving again, not only this PR. Peter and I discussed this last week.
> 
> First of all, since this PR doesn't break existing functionality, the general guidance about code moving in a good direction being mergeable applies.
> 

I think we can make this PR moving with minor changes.

> Second, one of the use cases for this level of logging is to help diagnose a failed run, at least on "replay" using the same seed and run configuration. In this case having a unique ID for each walker would be very useful, so this is what we should work towards. There is no question that we would have used the capability for this purpose several times in the last year.
> 

Indexing is unavoidable when recording into a big buffer/file. What I'm strongly against is to keep it as a state of walker. Indexing should be short lived whenever possible. Each element should avoid remembering its index, only the object handling a set of elements deal with indexing in restricted scopes.

> In VMC the IDs are trivial, since the walker count is constant. In DMC we need e.g. an incrementing counter and to make sure we can count 100s millions of walkers. Since the walker creation/destruction logic is global this is also not difficult. A new ID is only needed when new walkers are created. The required state information is a single (long?) integer. This can be worked towards in subsequent PRs.
> 
Creating new ID means the ID of other walkers can be out-dated. This requires global synchronization.

> As with general use of the per particle logger, we are concerned with functionality and achievable insight, not straight line speed.

**Comment by prckent** [comment]
> Creating new ID means the ID of other walkers can be out-dated. This requires global synchronization.

The ideas is that walkers have unique IDs for their entire lifetime.
If the ID is generated based on uniquely incrementing counter(s) there is no issue with them becoming out of date.
We already have a global sync to decide if to destroy or create walkers, so the IDs can be set or inferred then.
Something to work towards.

**Comment by ye-luo** [comment]
I still need an answer for how walker id is used for in this PR and why this is necessity. I don't mean to block this PR on this but the current use needs to be documented. How do we want that in the future will be a separate topic.

**Review by ye-luo — DISMISSED** [review]
Note: there is still no documentation about walker ID in this PR.

**Comment by ye-luo** [comment]
Test this please

**Comment by PDoakORNL** [comment]
test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**